### ISSUE TITLE
Add support for YAML streams from other sources

### ISSUE THREAD

**Issue by gitfool** [andrewlock/NetEscapades.Configuration#46]
@andrewlock Following up from https://github.com/wintoncode/Winton.Extensions.Configuration.Consul/pull/81#issuecomment-542407150, where I'd like to leverage your library's YAML parser after getting config from another source; Consul in this case. Thoughts?

**Comment by gitfool** [comment]
The only way I could easily get this to work was to lift the implementation of [YamlConfigurationFileParser](https://github.com/andrewlock/NetEscapades.Configuration/blob/master/src/NetEscapades.Configuration.Yaml/YamlConfigurationFileParser.cs#L10) and mix-in [IConfigurationParser](https://github.com/wintoncode/Winton.Extensions.Configuration.Consul/blob/master/src/Winton.Extensions.Configuration.Consul/Parsers/IConfigurationParser.cs#L12). I don't suppose it could be made `public` instead of `internal`? 😛

**Comment by gitfool** [comment]
Interestingly, I'm seeing exceptions due to multi-threaded re-entrancy. In the end it's my problem but I think `YamlConfigurationFileParser` could be improved slightly to enforce safe use from background threads, especially if it's to be exposed.

[Parse](https://github.com/andrewlock/NetEscapades.Configuration/blob/master/src/NetEscapades.Configuration.Yaml/YamlConfigurationFileParser.cs#L16) should be `static` and delegate to a `private ParseStream` on a `new` instance:
```c#
public static IDictionary<string, string> Parse(Stream input)
    => new YamlConfigurationFileParser().ParseStream(input);

private IDictionary<string, string> ParseStream(Stream input)
{
    _data.Clear();
    _context.Clear();
    ...
}
```
This would be similar to how [JsonConfigurationFileParser](https://github.com/aspnet/Extensions/blob/master/src/Configuration/Config.Json/src/JsonConfigurationFileParser.cs#L20-L23) is encapsulated, which your implementation follows. Also, given this pattern, the `Clear` calls on `_data` and `_context` are redundant and can be removed.

**Comment by goldsam** [comment]
@andrewlock  I'm working on a PR to introduce an `AddYamlStream` extension method.  To achieve this, I will use [`StreamConfigurationProvider`](https://github.com/dotnet/runtime/blob/main/src/libraries/Microsoft.Extensions.Configuration/src/StreamConfigurationProvider.cs) and mirror the implementation of [`JsonStreamConfigurationProvider`](https://github.com/dotnet/runtime/blob/215b39abf947da7a40b0cb137eab4bceb24ad3e3/src/libraries/Microsoft.Extensions.Configuration.Json/src/JsonStreamConfigurationProvider.cs). This requires updating to a newer `Microsoft.Extensions.Configuration` dependency which would break compatibility with Dot NET 2.x and Framework 4.6.1 which are both now out of support.

I noticed you were still maintaining support for Dot NET 2.x which was out of support at the time of your last 2.2.0 release. Is there a reason you are trying to maintain support for these legacy frameworks? Dropping support for core 2.x and 3.x would simplify changes and avoid the need for framework-specific conditional code compilation. Are you OK with dropping 4.6.1 in favor of 4.6.2 and dropping support for Core 2.x and 3.x  in favor of net6.0 and net7.0? This would be consistent with the most recent release of `Microsoft.Extensions.Configuration`.

**Comment by andrewlock** [comment]
Sorry for the delay, added an implementation in #76 and #77. It will be available in 3.1.0
